packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -837,6 +860,181 @@ function rerenderReducer<S, I, A>(
   return [newState, dispatch];
 }
 
+type MutableSourceState<Source, Snapshot> = {|
+  config: MutableSourceHookConfig<Source, Snapshot>,
+  destroy?: () => void,
+  snapshot: Snapshot,
+  source: MutableSource<any>,
+|};
+
+function useMutableSourceImpl<Source, Snapshot>(
+  hook: Hook,
+  source: MutableSource<Source>,
+  config: MutableSourceHookConfig<Source, Snapshot>,
+): Snapshot {
+  const {getSnapshot, subscribe} = config;
+  const version = source._getVersion();
+  const pendingExpirationTime = getPendingExpirationTime(
+    currentlyRenderingRoot,
+    source,
+  );
+
+  // Is it safe to read from this source during the current render?
+  let isSafeToReadFromSource = false;
+
+  if (pendingExpirationTime !== null) {
+    // If the source has pending updates, we can use the current render's expiration
+    // time to determine if it's safe to read again from the source.
+    const currentTime = requestCurrentTimeForUpdate();
+    const suspenseConfig = requestCurrentSuspenseConfig();
+    const expirationTime = computeExpirationForFiber(
+      currentTime,
+      currentlyRenderingFiber,
+      suspenseConfig,
+    );
+
+    isSafeToReadFromSource =
+      pendingExpirationTime === NoWork ||
+      pendingExpirationTime >= expirationTime;
```

Comment:
**Note to self**: If this component is newly mounting, we need to always check the version number as well, to account for the case of tearing between newly mounted components that have not yet subscribed to the store (and so may miss a tear).

I'll add a test for this (and update the code) tomorrow.

Reply:
Yes, this sounds right. I could avoid checking the expiration time if I already have a WIP version. Good suggestion.